From 1d94684ef5c90afa71ec52e8a9d1c33ba24c2f83 Mon Sep 17 00:00:00 2001
From: Nicholas Piggin <npiggin@gmail.com>
Date: Mon, 5 Feb 2024 23:02:08 +1000
Subject: [PATCH] ui: Fix compile with curses opaque menu headers

Recent Debian ncurses is built with NCURSES_OPAQUE_MENU, which means
the menu item type can't be poked directly. Fix with accessors.

One wrinkle is the item name can't be modified, new_item has to be used.

Signed-off-by: Nicholas Piggin <npiggin@gmail.com>
Signed-off-by: Jeremy Kerr <jk@codeconstruct.com.au>
---
 ui/ncurses/nc-menu.c | 12 ++++++++----
 ui/ncurses/nc-menu.h |  4 ++--
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/ui/ncurses/nc-menu.c b/ui/ncurses/nc-menu.c
index a90a02e3..a1524b5c 100644
--- a/ui/ncurses/nc-menu.c
+++ b/ui/ncurses/nc-menu.c
@@ -141,9 +141,13 @@ int pmenu_item_update(struct pmenu_item *item, const char *name)
 	if (!label)
 		return -1;
 
-	i = item->nci;
-	i->name.str = label;
-	i->name.length = strncols(label);
+	i = new_item(label, NULL);
+	if (!i) {
+		talloc_free((char *)label);
+		return -1;
+	}
+	free_item(item->nci);
+	item->nci = i;
 
 	return 0;
 }
@@ -358,7 +362,7 @@ static int pmenu_item_get_index(const struct pmenu_item *item)
 				return i;
 
 	pb_log_fn("not found: %p %s\n", item,
-		(item ? item->nci->name.str : "(null)"));
+		(item ? item_name(item->nci) : "(null)"));
 	return -1;
 }
 
diff --git a/ui/ncurses/nc-menu.h b/ui/ncurses/nc-menu.h
index eb568c87..550c7e13 100644
--- a/ui/ncurses/nc-menu.h
+++ b/ui/ncurses/nc-menu.h
@@ -126,7 +126,7 @@ static inline struct pmenu *pmenu_from_scr(struct nc_scr *scr)
 
 static inline void pmenu_dump_item(const ITEM *item)
 {
-	pb_debug("%p %s\n", item, (item ? item->name.str : "(null)"));
+	pb_debug("%p %s\n", item, (item ? item_name(item) : "(null)"));
 }
 
 static inline void pmenu_dump_items(ITEM *const *items, unsigned int count)
@@ -135,7 +135,7 @@ static inline void pmenu_dump_items(ITEM *const *items, unsigned int count)
 
 	for (i = 0; i < count; i++)
 		pb_debug("%u: %p %s\n", i, items[i],
-			(items[i] ? items[i]->name.str : "(null)"));
+			(items[i] ? item_name(items[i]) : "(null)"));
 }
 
 #endif
